src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLSurfaceDataBase.h

```diff
@@ -0,0 +1,150 @@
+ * The offset in pixels of the Metal viewport origin from the lower-left
+ * corner of the heavyweight drawable.  For example, a top-level frame on
+ * Windows XP has lower-left insets of (4,4).  The Metal viewport origin
+ * would typically begin at the lower-left corner of the client region (inside
+ * the frame decorations), but AWT/Swing will take the insets into account
+ * when rendering into that window.  So in order to account for this, we
+ * need to adjust the Metal viewport origin by an x/yOffset of (-4,-4).  On
+ * X11, top-level frames typically don't have this insets issue, so their
+ * x/yOffset would be (0,0) (the same applies to pbuffers).
+ *
+ *     jint width/height;
+ * The cached surface bounds.  For offscreen surface types (MTLSD_FBOBJECT,
+ * MTLSD_TEXTURE, etc.) these values must remain constant.  Onscreen window
+ * surfaces (MTLSD_WINDOW, MTLSD_FLIP_BACKBUFFER, etc.) may have their
+ * bounds changed in response to a programmatic or user-initiated event, so
+ * these values represent the last known dimensions.  To determine the true
+ * current bounds of this surface, query the native Drawable through the
+ * privOps field.
+ *
+ *     void* pTexture;
+ * The texture object handle, as generated by MTLTextureDescriptor(). If this
+ * value is null, the texture has not yet been initialized.
+ *
+ *     jint textureWidth/Height;
+ * The actual bounds of the texture object for this surface.
+ * The texture image that we care about has dimensions specified by the width
+ * and height fields in this MTLSDOps structure.
+ */
+typedef struct {
+    SurfaceDataOps               sdOps;
+    void                         *privOps;
+    jobject                      graphicsConfig;
+    jint                         drawableType;
+    jboolean                     isOpaque;
+    jint                         xOffset;
+    jint                         yOffset;
+    jint                         width;
+    jint                         height;
+    void*                        pTexture;
+    void*                        pStencilData;      // stencil data to be rendered to this buffer
+    void*                        pStencilDataBuf;   // MTLBuffer with stencil data
+    void*                        pStencilTexture;   // stencil texture byte buffer stencil mask used in main rendering
+    void*                        pAAStencilData;    // stencil data for AA rendering
+    void*                        pAAStencilDataBuf; // MTLBuffer with AA stencil data
+    jint                         textureWidth;
+    jint                         textureHeight;
+} BMTLSDOps;
+
+#define MTLSD_UNDEFINED       sun_java2d_pipe_hw_AccelSurface_UNDEFINED
+#define MTLSD_WINDOW          sun_java2d_pipe_hw_AccelSurface_WINDOW
+#define MTLSD_TEXTURE         sun_java2d_pipe_hw_AccelSurface_TEXTURE
+#define MTLSD_FLIP_BACKBUFFER sun_java2d_pipe_hw_AccelSurface_FLIP_BACKBUFFER
+#define MTLSD_RT_TEXTURE        sun_java2d_pipe_hw_AccelSurface_RT_TEXTURE
```

Comment:
Only two of them are supported? MTLSD_TEXTURE and MTLSD_RT_TEXTURE?

Reply:
Also, MTLSD_WINDOW is supported

Reply:
How it is used? Can we really draw to the window directly? It looks like it copied from the OGL where it is unused since CLayer integration. Something is inconsistent here.

Reply:
Here is a follow-up issue JDK-8263463